clash-prelude/src/Clash/Class/Resize.hs

```diff
@@ -117,3 +129,38 @@ checkedTruncateB ::
   f (a + b) -> f a
 checkedTruncateB v =
   checkIntegral (Proxy @(f a)) v `seq` truncateB v
+
+-- | Like 'resize', but returns 'Nothing' if the argument is out of bounds for
+-- the result type.
+maybeResize ::
+  forall a b f.
+  ( Resize f
+  , KnownNat a, Integral (f a)
+  , KnownNat b, Integral (f b), Bounded (f b) ) =>
+  f a -> Maybe (f b)
+maybeResize v =
+#if MIN_VERSION_base(4,16,0)
+  case Proxy @a `cmpNat` Proxy @b of
+    GTI | v > resize (maxBound @(f b)) -> Nothing
+    GTI | v < resize (minBound @(f b)) -> Nothing
+    EQI -> Just v
+    _ -> Just (resize v)
+#else
+  case natToNum @a @Int `compare` natToNum @b @Int of
```

Comment:
What about `Index (2^65)`? (Do we still do 32-bit machine compatibility? If so, what about `Index (2^33)`?)

I feel it would be better to use `Integer` or `Natural` here.

Reply:
```haskell
case natToNatural @a `compare` natToNatural @b of
```

Reply:
Oh wow, I didn't even consider that. I guess the tests should pick from `[1..2^128]` too!